src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -15,20 +13,23 @@ internal sealed class ServiceProviderEngineScope : IServiceScope, IServiceProvid
         // For testing only
         internal Action<object> _captureDisposableCallback;
 
-        private List<object> _disposables;
-
         private bool _disposed;
-        private readonly object _disposelock = new object();
+        private ScopePool.State _state;
+
+        // This lock protects state on the scope, in particular, for the root scope, it protects
+        // the list of disposable entries only as ResolvedServices is a concurrent dictionary
+        // For other scopes, it protects ResolvedServices and the list of disposables
+        private readonly object _scopeLock = new object();
 
         public ServiceProviderEngineScope(ServiceProviderEngine engine, bool isRoot = false)
         {
             Engine = engine;
-
-            // To reduce lock contention for singletons upon resolve we use a concurrent dictionary.
-            ResolvedServices = isRoot ? new ConcurrentDictionary<ServiceCacheKey, object>() : new Dictionary<ServiceCacheKey, object>();
+            _state = isRoot ? new ScopePool.State() : engine.ScopePool.Rent();
         }
 
-        internal IDictionary<ServiceCacheKey, object> ResolvedServices { get; }
+        internal IDictionary<ServiceCacheKey, object> ResolvedServices => _state?.ResolvedServices ?? ScopeDisposed();
```

Comment:
would it be easy to add tests for some of these cases? like when `ScopeDisposed` would happen

Reply:
There's one test that was modified to show this actually the only modified test